query-engine/core/src/query_graph/mod.rs

```diff
@@ -831,6 +838,108 @@ impl QueryGraph {
+    ///     (A, B)                (B, C)           (A, D)
+    ///        │                    │               │
+    ///        ▼                    ▼               ▼
+    /// ┌ ─ ─ ─ ─ ─ ─          ┌ ─ ─ ─ ─ ─ ─   ┌ ─ ─ ─ ─ ─ ─
+    ///    Child A   │         |  Child B   │  |  Child C   │
+    /// └ ─ ─ ─ ─ ─ ─          └ ─ ─ ─ ─ ─ ─   └ ─ ─ ─ ─ ─ ─
+    /// ```
+    /// However, `Parent` only returns `(A, B)`, for example, because that's the primary ID of the parent model.
+    ///
+    /// In order to satisfy children B and C, the graph is altered by this post-processing call:
+    /// ```text
+    /// ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+    ///     Parent (A, B, C, D)  │─────────┬───────────────┐
+    /// └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─          │               │
+    ///        │                           │               │
+    ///     (A, B)                       (B, C)           (A, D)
+    ///        │                           │               │
+    ///        ▼                           ▼               ▼
+    /// ┌ ─ ─ ─ ─ ─ ─               ┌ ─ ─ ─ ─ ─ ┐   ┌ ─ ─ ─ ─ ─ ┐
+    ///    Child A   │              |  Child B  │   |  Child C  │
+    /// └ ─ ─ ─ ─ ─ ─               └ ─ ─ ─ ── -    └ ─ ─ ─ ── ─
+    /// ```
+    /// Note that not all connectors can have their nodes' field selection updated.
+    /// This is only possible when the parent node _can_ fulfill the selection set.
+    /// In the case of updates and inserts, for instance, only connectors supporting `InsertReturning` and `UpdateReturning` can do it,
+    /// or else they're only able to return the primary identifier of the model inserted or updated.
+    fn normalize_data_dependencies(&mut self, capabilities: ConnectorCapabilities) -> QueryGraphResult<()> {
+        let unsatisfied_deps: Vec<(NodeRef, Vec<FieldSelection>)> = self
+            .graph
+            .node_indices()
+            .filter_map(|ix| -> Option<(NodeRef, Vec<FieldSelection>)> {
+                let node = NodeRef { node_ix: ix };
+
+                if let Node::Query(q) = self.node_content(&node).unwrap() {
+                    // If the connector does not support returning more than the primary identifier for an update,
+                    // do not update the selection set.
+                    if q.is_update_one() && !capabilities.contains(ConnectorCapability::UpdateReturning) {
+                        return None;
+                    }
+
+                    // If the connector does not support returning more than the primary identifier for a create,
+                    // do not update the selection set.
+                    if q.is_create_one() && !capabilities.contains(ConnectorCapability::InsertReturning) {
+                        return None;
+                    }
+
+                    // No connector supports returning more than the primary identifier for a delete just yet.
+                    if q.is_delete_one() {
+                        return None;
+                    }
```

Comment:
This is needed to ensure we don't update the selection set of nodes that eventually won't be fulfilled.

For PG & CRDB on update & create nodes, that means the `insert_reloads` won't touch these nodes on the second pass.

For other connectors, that means most nodes won't be touched on this pass but the second pass (`insert_reloads`) will add the relevant reload nodes.

I intentionally did not merge the two functions together to let them act as independent passes on the graph.